--- a/libavformat/libsmbclient.c
+++ b/libavformat/libsmbclient.c
@@ -33,6 +33,8 @@
     int trunc;
     int timeout;
     char *workgroup;
+    char *username;
+    char *password;
 } LIBSMBContext;
 
 static void libsmbc_get_auth_data(SMBCCTX *c, const char *server, const char *share,
@@ -40,8 +42,14 @@
                                   char *username, int username_len,
                                   char *password, int password_len)
 {
-    /* Do nothing yet. Credentials are passed via url.
-     * Callback must exists, there might be a segmentation fault otherwise. */
+    LIBSMBContext *libsmbc = smbc_getOptionUserData(c);
+
+    if (libsmbc->workgroup)
+        av_strlcpy(workgroup, libsmbc->workgroup, workgroup_len);
+    if (libsmbc->username)
+        av_strlcpy(username, libsmbc->username, username_len);
+    if (libsmbc->password)
+        av_strlcpy(password, libsmbc->password, password_len);
 }
 
 static av_cold int libsmbc_connect(URLContext *h)
@@ -61,7 +69,7 @@
     }
     smbc_set_context(libsmbc->ctx);
 
-    smbc_setOptionUserData(libsmbc->ctx, h);
+    smbc_setOptionUserData(libsmbc->ctx, libsmbc);
     smbc_setFunctionAuthDataWithContext(libsmbc->ctx, libsmbc_get_auth_data);
 
     if (libsmbc->timeout != -1)
@@ -351,6 +359,8 @@
 #define D AV_OPT_FLAG_DECODING_PARAM
 #define E AV_OPT_FLAG_ENCODING_PARAM
 static const AVOption options[] = {
+    {"username", "set the username used for SMB authentication", OFFSET(username), AV_OPT_TYPE_STRING, { 0 }, 0, 0, D|E },
+    {"password", "set the password used for SMB authentication", OFFSET(password), AV_OPT_TYPE_STRING, { 0 }, 0, 0, D|E },
     {"timeout",   "set timeout in ms of socket I/O operations",    OFFSET(timeout), AV_OPT_TYPE_INT, {.i64 = -1}, -1, INT_MAX, D|E },
     {"truncate",  "truncate existing files on write",              OFFSET(trunc),   AV_OPT_TYPE_INT, { .i64 = 1 }, 0, 1, E },
     {"workgroup", "set the workgroup used for making connections", OFFSET(workgroup), AV_OPT_TYPE_STRING, { 0 }, 0, 0, D|E },
